Support cursor visibility based on input device type

freemangordon opened the pull request:
The logic is: if we receive motion event from a TS device, hide the cursor,
show it otherwise.

Most of the code was borrowed from Cordia h-d

Signed-off-by: Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>

spinal84 commented:
Hi. What compiler errors were produced?

freemangordon commented:
merged to maemo/ascii-devel, thanks

MerlijnWajer commented:
BTW: build went ok: https://phoenix.maemo.org/job/hildon-desktop-binaries/2/

MerlijnWajer commented on src/main.c:
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
+          xi_dev->is_ts = (info.type == atom_touchscreen);
+          xi_dev->dev = dev;
+#if 0
+          g_warning ("### %d/%u. %s(%d):%s.%d t:%d c:%d\n", i, eclass->len,
```
Don't care much, but did you want to keep this here, in '#if 0' ?

freemangordon replied:
It is very useful for debugging

spinal84 commented on src/main.c:
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
+      xi_devices = NULL;
+    }
+
+  if (atom_touchscreen == None)
+    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
+
+  if (atom_touchscreen == None)
+    return;
```
May be replaced with:
```
  if (atom_touchscreen == None) {
    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);

    if (atom_touchscreen == None)
      return;
  }
```

If we have `atom_touchscreen != None` (that is in all cases except the first call), we will just have two equal `if (...)` statements in a row, both of which will have FALSE.

freemangordon replied:
I don;t really like such nested ifs, that's why I do it like that. Regarding the performance - I can bet compiler will optimize.

spinal84 replied:
> Regarding the performance - I can bet compiler will optimize.

According to my tests with `gcc -O2 -S`, it generates additional assembler instruction for the case without nested if's (on amd64 arch). When gcc doesn't use -O2, the instruction set size is the same, but it executes the check inside the if statement for two times, just like it's done in the code.

Anyway, if you don't like nested 'if' so much, let's leave it intact.

spinal84 commented on src/main.c:
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
+  int i, ndev;
+  GArray *eclass;
+
+  if (xi_devices)
+    {
+      close_input_devices (dpy);
+      g_array_free (xi_devices, TRUE);
+      xi_devices = NULL;
+    }
```
I'd move this logic to `close_input_devices()` which may be renamed to `free_input_devices`.
So we just call `free_input_devices()` here and on hildon-desktop termination.

freemangordon replied:
Yes, it can be done as you say, but it is a matter of preference, There is no gain.

Also, freeing function a() allocated memory in function b() is bad coding practice, that's why it is done like that.

spinal84 commented on src/main.c:
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
+      if (xi_dev->dev)
+        {
+          XCloseDevice (dpy, xi_dev->dev);
+          xi_dev->dev = NULL;
```
Why is this needed?

freemangordon replied:
because it will prevent SEGFAULTs if somebody decides to (wrongly) call the function for the second time. Also, we better have no dangling pointers around.

spinal84 commented on src/main.c:
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
+    {
+      XDeviceInfo info = devinfo[i];
+
+      if (info.use == IsXExtensionPointer)
```
Shouldn't it be
`if (info.use == IsXPointer || info.use == IsXExtensionPointer)`?

freemangordon replied:
I think IsXPointer is for core pointer devices, i.e. those that are not a real mice or TSes. Do you think IsXPointer shall be included and why?

spinal84 replied:
Isn't it possible that the user configure a touchscreen as a core pointer device?

spinal84 replied:
> i.e. those that are not a real mice or TSes.

Could you please share a link to documentation about this?
I wasn't aware of it (core pointer devices to be virtual) and couldn't fine an information.

spinal84 replied:
From xorg.conf (5) manpage:

`If hotplugging is disabled, there will normally be at least two: one for the core (primary) keyboard and one for the core pointer. If either of these two is missing, a default configuration for the missing ones will be used. In the absence of an explicitly specified core input device, the first InputDevice marked as CorePointer (or CoreKeyboard) is used.`

freemangordon replied:
The point is that we wan't to make a difference whether a device is mouse or ts. Core pointers' type atom is null, so it makes no sense checking them. So, if we have core pointers only, cursor will be hidden initially and that's it.

spinal84 commented on src/main.c:
```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
 
+/* Close all input devices previously opened */
+static void
+close_input_devices (Display *dpy)
```
The call to this function is missing on hildon-desktop termination.

freemangordon replied:
Yes, but there is no graceful termination of hildon-desktop I was able to find.

spinal84 replied:
What about the code block coming after the call to `gtk_main()` in the main function of main.c file?

freemangordon replied:
ok, I'll put a call to close_input_devices() there, but I doubt it will ever be reached.

freemangordon committed:
Support cursor visibility based on input device type

The logic is: if we receive motion event from a TS device, hide the cursor,
show it otherwise.

Most of the code was borrowed from Cordia h-d

Signed-off-by: Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>
